### PR TITLE
[DT 1111]Added ud client as proxy server and remove default rpc url

### PR DESCRIPTION
**Pull request by tunguyennnnn**
## Background

This release offers a new interface for resolution library to call unstoppable domains proxy service for resolving onchain data.

## Changes

* Add new interface to configure interacting with proxy service
* Remove the use of default Infura and Alchemy keys
  * Migrate to use environment variables   for unit test
  * Remove configuration of default RPC provider

## Code Review

This Pull Request was thoroughly reviewed and meets all Code Review Standards pertaining to:

- [ ] **Implementation** - satisfied acceptance criteria
- [ ] **Communication** - notified engineers/stakeholders about impactful changes
- [ ] **Error handling** - handled, logged & alerted on errors
- [ ] **Documentation** - wrote readable code & commits, documented unintuitive code
- [ ] **Testing** - unit & E2E test coverage, tested in staging, DB migrations backwards compatible
- [ ] **Security** - sanitized inputs, vetted dependencies, validated/secured API requests, no committed secrets
- [ ] **Performance** - optimized expensive algorithms & database queries

Please select all **applied** standards relevant to this PR.
## Confirmation

- [ ] **Assignee Confirmation** - I (the author) have filled out the template above
- [ ] **Reviewer Confirmation** - I (a/the reviewer) confirm 1) the author has filled out the template and checked the box that says **Assignee Confirmation** 2) I reviewed the code and selected all applicable items in the code review section.

### CODE REVIEW

File: src/main/java/com/unstoppabledomains/exceptions/ns/NamingServiceException.java [hunk 1]
```diff
@@ -61,6 +61,9 @@ private static String messageFromCode(NSExceptionCode code, NSExceptionParams pa
       case ReverseResolutionNotSpecified: {
         return "Reverse resolution is not specified for address " + params.contractAddress;
       }
+      case RPCServerError: {
+        return "RPC server error " + params.serverMessage;
```
Comment (sammyluo) [inline src/main/java/com/unstoppabledomains/exceptions/ns/NamingServiceException.java:65]: Nit:
```suggestion
        return "RPC server error: " + params.serverMessage;
```

File: src/main/java/com/unstoppabledomains/resolution/Resolution.java [hunk 2]
```diff
@@ -261,13 +261,16 @@ private TokenUriMetadata getMetadataFromTokenURI(String tokenURI) throws NamingS
         }
     }
 
-    private Map<NamingServiceType, NamingService> getServices(IProvider provider) {
+    private Map<NamingServiceType, NamingService> getServices(IProvider provider, String apiKey) {
+        provider.setHeader("Authorization", "Bearer " + apiKey);
+        provider.setHeader("X-Lib-Client", DefaultProvider.getUserAgent());
```
Comment (sammyluo) [inline src/main/java/com/unstoppabledomains/resolution/Resolution.java:266]:
```suggestion
        provider.setHeader("X-Lib-Agent", DefaultProvider.getUserAgent());
```

File: src/main/java/com/unstoppabledomains/resolution/contracts/interfaces/IProvider.java [hunk 3]
```diff
@@ -13,4 +13,12 @@ public interface IProvider {
    * @throws IOException
    */
   abstract JsonObject request(String url, JsonObject body) throws IOException;
+
+    /**
```
Comment (sammyluo) [inline src/main/java/com/unstoppabledomains/resolution/contracts/interfaces/IProvider.java:17]: Nit: indentation
```suggestion
   /**
```

File: src/main/java/com/unstoppabledomains/resolution/contracts/BaseContract.java [hunk 4]
```diff
@@ -99,6 +99,13 @@ private Tuple fetchMethod(String method, Object[] args) throws NamingServiceExce
       final String replacedAnswer = answer.replace("0x", "");
       return function.decodeReturn(FastHex.decode(replacedAnswer));
     } catch(IOException exception) {
+      if (exception.getMessage().matches("(.*)response code: (403|401|429)(.*)")) {
```
Comment (sammyluo) [inline src/main/java/com/unstoppabledomains/resolution/contracts/BaseContract.java:102]: What about 5xx?

**Review by sammyluo — CHANGES_REQUESTED** [review]
Good start!

File: src/main/java/com/unstoppabledomains/resolution/contracts/BaseContract.java (hunk 4 above)
Comment (tunguyennnnn) [reply src/main/java/com/unstoppabledomains/resolution/contracts/BaseContract.java:102]: these are the error code that will make partners to be aware of when integrate with us such as authorized and rate limited. The others like 5xx could be thrown by RPC providers as well so let them handle it themselves.

**Review by sammyluo — APPROVED** [review]
Awesome! 🚀 ⭐

**Pull request merged**